src/cargo/util/toml/targets.rs

```diff
@@ -719,16 +720,26 @@ fn inferred_to_toml_targets(inferred: &[(String, PathBuf)]) -> Vec<TomlTarget> {
         .collect()
 }
 
-fn validate_has_name(
+fn validate_target_name(
     target: &TomlTarget,
     target_kind_human: &str,
     target_kind: &str,
+    warnings: &mut Vec<String>,
 ) -> CargoResult<()> {
     match target.name {
         Some(ref name) => {
             if name.trim().is_empty() {
                 anyhow::bail!("{} target names cannot be empty", target_kind_human)
             }
+            if cfg!(windows) {
+                if restricted_names::is_windows_reserved(name) {
+                    warnings.push(format!(
+                        "{} target `{}` is a reserved Windows filename, \
+                        this target will not work on Windows platforms",
+                        target_kind_human, name
+                    ));
```

Comment:
Hm can you clarify this a bit for me? How come we'd only issue this warning on windows, and how comit it applies to all targets? It should be ok to have a library called `con`, right? I don't think that creates any of the bad file names?

Reply:
My thinking for only showing on Windows is:  if someone wants to create an executable called `con`, and they never intend for it to be used on Windows, then that would create a warning that they cannot silence.  Maybe if we had the "compatibility" table we discussed at the last meeting, we could warn on all platforms and then allow the warning to be silenced by specifying that it is incompatible with windows?

FWIW, I scanned all of crates.io, and didn't find any targets with a reserved name.  Part of this came up with a user on Discord who had a package named `prune` with a binary named `prn`, and was confused when their Windows CI failed.  I don't think they've published to crates.io, though.

As for libraries, in some situations rustc creates filenames like `con.tr1cof326ip8vwg.rcgu.o` which fail on windows.  It also seems like it would just be a risky thing to do, and I didn't want to put in effort and complexity to handle that edge case, which is pretty far on the edge.

Reply:
Oh for sure binaries make sense here, I was just confused about the library part. The point about crate names showing up in object files makes sense though!